extensions/notebook/src/test/model/jupyterController.test.ts

```diff
@@ -0,0 +1,53 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import * as should from 'should';
+import * as TypeMoq from 'typemoq';
+
+import { ApiWrapper } from '../../common/apiWrapper';
+import { AppContext } from '../../common/appContext';
+import { JupyterController } from '../../jupyter/jupyterController';
+import { LocalPipPackageManageProvider } from '../../jupyter/localPipPackageManageProvider';
+import { MockExtensionContext } from '../common/stubs';
+
+describe('JupyterController tests', function () {
+	let mockExtensionContext: vscode.ExtensionContext;
+	let appContext: AppContext;
+	let controller: JupyterController;
+	let mockApiWrapper: TypeMoq.IMock<ApiWrapper>;
+
+	this.beforeAll(async () => {
+		mockExtensionContext = new MockExtensionContext();
+		mockApiWrapper = TypeMoq.Mock.ofType<ApiWrapper>();
+		appContext = new AppContext(mockExtensionContext, mockApiWrapper.object);
+	});
+
+	this.beforeEach(async () => {
+		controller = new JupyterController(appContext);
+	});
+
+	it('should activate new JupyterController successfully', async () => {
+		should(controller.extensionContext).deepEqual(appContext.extensionContext, 'Extension context should be passed through');
+		should(await controller.activate()).not.throw();
```

Comment:
This doesn't work like that - for promises you have use rejected.

`await should(controller.activate()).not.be.rejected()`

The not throw doesn't handle async code so if an error was thrown it would escape the check.

Now in this particular case it doesn't matter since something throwing unexpectedly is just going to fail the test regardless. So I'd just say you can remove the not.throw check completely. The above is more useful when you need to test that something DOES throw.

`await should(controller.activate()).be.rejected()`;

Reply:
Good call. Before I mocked out the ApiWrapper, this was actually throwing previously, so this is just a result of that.

Let me change this, appreciate the thoroughness here.